elastic/translate.go

```diff
@@ -1,185 +1,186 @@
 package elastic
 
 import (
-	"encoding/json"
+	"errors"
 	"fmt"
-	"log"
 	"strings"
 )
 
 // Leaf is a terminal node of a query
 // All terminal query terms implement the Translate() method
-type Leaf interface {
-	Translate() map[string]interface{}
+type leaf interface {
+	Translate() Term
 }
 
-func encodeParam(concept string, param map[string]interface{}) map[string]interface{} {
-	operator := param["operator"].(string)
-	var conceptID string
-	if operator == "set" {
-		conceptID = "_id"
+// encodeParam encodes a parameter from a query to its corresponding leaf type
+func encodeParam(concept string, param map[string]interface{}) (map[string]interface{}, error) {
+	o := param["operator"].(string)
+	var cid string
+	if o == "set" {
+		cid = "_id"
 	} else {
-		conceptID = param["id"].(string)
+		cid = param["id"].(string)
 	}
 	var conceptPath string
-	if concept != "" && operator != "set" {
-		conceptPath = concept + "." + conceptID
+	if concept != "" && o != "set" {
+		conceptPath = concept + "." + cid
 	} else {
-		conceptPath = conceptID
+		conceptPath = cid
 	}
-	value := param["value"]
-	var paramLeaf Leaf
-	switch operator {
+	v := param["value"]
+	var l leaf
+	switch o {
 	case "set":
-		paramLeaf = &MemberTerm{conceptPath, value}
+		l = &MemberTerm{conceptPath, v}
 	case "eq":
-		paramLeaf = &EqualityTerm{conceptPath, value, "+"}
+		l = &EqualityTerm{conceptPath, v, false}
 	case "-eq":
-		paramLeaf = &EqualityTerm{conceptPath, value, "-"}
+		l = &EqualityTerm{conceptPath, v, true}
 	case "undefined":
-		paramLeaf = &DefinitionTerm{conceptPath, false}
+		l = &DefinitionTerm{conceptPath, false}
 	case "defined":
-		paramLeaf = &DefinitionTerm{conceptPath, true}
+		l = &DefinitionTerm{conceptPath, true}
 	case "one":
-		sliceValue := value.([]string)
-		paramLeaf = &OneTerm{conceptPath, sliceValue}
+		s := v.([]string)
```

Comment:
A bad type assertion will cause a panic. Do

```go
s, ok := v.([]string)
if !ok {
    return nil, errors.New("...")
}
```

Also, that assertion will always fail since the JSON decoder only put things in `[]interface{}`. I mentioned this before, but an alternate strategy could be to incrementally decode the JSON as you switch on various keywords. For example:

```go
// Given some valid JSON...
var v map[string]json.RawMessage
json.Unmarshal(b, &v)

for k, r := range v {
    switch k {
    case "match":
        // Requires fields defined on struct to not be private.
        var t MatchTerm
        json.Unmarshal(r, &t)
    }
}
```

See the example [in the docs](https://golang.org/pkg/encoding/json/#RawMessage) for a live preview.